File: benchmarking/bench-cluster/java/BenchCore.java
import java.util.*;
import java.util.concurrent.*;
import java.io.*;
import java.nio.file.*;
import com.statsig.*;
import com.alibaba.fastjson2.JSON;
import com.alibaba.fastjson2.JSONWriter;

public class BenchCore {
    static final String sdkType = "statsig-server-core-java";
    static final String SCRAPI_URL = "http://scrapi:8000";
    static final int ITER_LITE = 1000;
    static final int ITER_HEAVY = 10_000;
    static final Random random = new Random();

    public static void main(String[] args) throws Exception {
        String sdkVersion = getSdkVersion();
        System.out.println("Statsig Java Core (v" + sdkVersion + ")");
        System.out.println("--------------------------------");

        Map<String, List<String>> specNames = loadSpecNames();

        StatsigOptions options = new StatsigOptions.Builder()
            .setSpecsUrl(SCRAPI_URL + "/v2/download_config_specs")
            .setLogEventUrl(SCRAPI_URL + "/v1/log_event")
            .build();

        Statsig statsig = new Statsig("secret-JAVA_CORE", options);
        statsig.initialize().get();

        List<BenchmarkResult> results = new ArrayList<>();

        // Create a global user
        StatsigUser globalUser = new StatsigUser.Builder().setUserID("global_user").build();

        // Feature gates
        for (String gate : specNames.get("feature_gates")) {
            benchmark("check_gate", gate, ITER_HEAVY, () -> statsig.checkGate(createUser(), gate), results, sdkVersion);
            benchmark("check_gate_global_user", gate, ITER_HEAVY, () -> statsig.checkGate(globalUser, gate), results, sdkVersion);
            benchmark("get_feature_gate", gate, ITER_HEAVY, () -> statsig.getFeatureGate(createUser(), gate), results, sdkVersion);
            benchmark("get_feature_gate_global_user", gate, ITER_HEAVY, () -> statsig.getFeatureGate(globalUser, gate), results, sdkVersion);
        }

        // Dynamic configs
        for (String config : specNames.get("dynamic_configs")) {
            benchmark("get_dynamic_config", config, ITER_HEAVY, () -> statsig.getDynamicConfig(createUser(), config), results, sdkVersion);
            benchmark("get_dynamic_config_global_user", config, ITER_HEAVY, () -> statsig.getDynamicConfig(globalUser, config), results, sdkVersion);
        }

        // Experiments
        for (String experiment : specNames.get("experiments")) {
            benchmark("get_experiment", experiment, ITER_HEAVY, () -> statsig.getExperiment(createUser(), experiment), results, sdkVersion);
            benchmark("get_experiment_global_user", experiment, ITER_HEAVY, () -> statsig.getExperiment(globalUser, experiment), results, sdkVersion);
        }

        // Layers
        for (String layer : specNames.get("layers")) {
            benchmark("get_layer", layer, ITER_HEAVY, () -> statsig.getLayer(createUser(), layer), results, sdkVersion);
            benchmark("get_layer_global_user", layer, ITER_HEAVY, () -> statsig.getLayer(globalUser, layer), results, sdkVersion);
        }

        // Client initialize response
        benchmark("get_client_initialize_response", "n/a", ITER_LITE, () -> statsig.getClientInitializeResponse(createUser()), results, sdkVersion);
        benchmark("get_client_initialize_response_global_user", "n/a", ITER_LITE, () -> statsig.getClientInitializeResponse(globalUser), results, sdkVersion);

        statsig.shutdown().get();

        // Write results
        writeResults(sdkType, sdkVersion, results);
    }

    private static Map<String, List<String>> loadSpecNames() throws Exception {
        String path = "/shared-volume/spec_names.json";
        for (int i = 0; i < 10; i++) {
            if (Files.exists(Paths.get(path))) break;
            Thread.sleep(1000);
        }
        String json = Files.readString(Paths.get(path));
        return JSON.parseObject(json, Map.class);
    }

    private static StatsigUser createUser() {
        return new StatsigUser.Builder()
            .setUserID("user_" + random.nextInt(1000000))
            .build();
    }


    // --- Benchmarking logic ---
    private static void benchmark(String benchName, String specName, int iterations, Runnable func, List<BenchmarkResult> results, String sdkVersion) {
        List<Double> durations = new ArrayList<>();
        for (int i = 0; i < iterations; i++) {
            long start = System.nanoTime();
            func.run();
            long end = System.nanoTime();
            durations.add((end - start) / 1_000_000.0); // ms
        }
        Collections.sort(durations);
        int p99Index = (int)(iterations * 0.99);
        BenchmarkResult result = new BenchmarkResult(
            benchName,
            durations.get(p99Index),
            durations.get(durations.size() - 1),
            durations.get(0),
            durations.get(durations.size() / 2),
            durations.stream().mapToDouble(Double::doubleValue).average().orElse(0.0),
            specName,
            sdkType,
            sdkVersion
        );
        results.add(result);
        System.out.printf("%-30s p99(%.4fms) max(%.4fms) %s\n", benchName, result.p99, result.max, specName);
        try { Thread.sleep(1); } catch (InterruptedException ignored) {}
    }

    private static void writeResults(String sdkType, String sdkVersion, List<BenchmarkResult> results) throws Exception {
        HashMap<String, Object> root = new HashMap<>();
        root.put("sdkType", sdkType);
        root.put("sdkVersion", sdkVersion);
        root.put("results", results);
        String outPath = String.format("/shared-volume/%s-%s-results.json", sdkType, sdkVersion);
        Files.writeString(Paths.get(outPath), JSON.toJSONString(root, JSONWriter.Feature.PrettyFormat));
    }

    private static String getSdkVersion() throws Exception {
        String root = System.getProperty("user.dir");
        Properties props = new Properties();
        props.load(new FileInputStream(root + "/build/versions.properties"));
        String version = props.getProperty("core.version", "unknown");
        return version;
    }

    // --- BenchmarkResult class ---
    static class BenchmarkResult {
        public String benchmarkName;
        public double p99;
        public double max;
        public double min;
        public double median;
        public double avg;
        public String specName;
        public String sdkType;
        public String sdkVersion;

        public BenchmarkResult(String benchmarkName, double p99, double max, double min, double median, double avg, String specName, String sdkType, String sdkVersion) {
            this.benchmarkName = benchmarkName;
            this.p99 = p99;
            this.max = max;
            this.min = min;
            this.median = median;
            this.avg = avg;
            this.specName = specName;
            this.sdkType = sdkType;
            this.sdkVersion = sdkVersion;
        }
    }
}

File: benchmarking/bench-cluster/dotnet/BenchCore.cs
extern alias StatsigCore;

using System.Reflection;
using System.Diagnostics;
using System.Text.Json;
using StatsigCore::Statsig;
using System.Text.Json.Serialization;

namespace bench;


public class BenchCore
{
    public class BenchmarkResult
    {
        [JsonPropertyName("benchmarkName")]
        public string BenchmarkName { get; set; }
        [JsonPropertyName("p99")]
        public double P99 { get; set; }
        [JsonPropertyName("max")]
        public double Max { get; set; }
        [JsonPropertyName("min")]
        public double Min { get; set; }
        [JsonPropertyName("median")]
        public double Median { get; set; }
        [JsonPropertyName("avg")]
        public double Avg { get; set; }
        [JsonPropertyName("specName")]
        public string SpecName { get; set; }
        [JsonPropertyName("sdkType")]
        public string SdkType { get; set; }
        [JsonPropertyName("sdkVersion")]
        public string SdkVersion { get; set; }
    }

    static readonly string sdkVersion = GetVersion();
    static readonly Random rng = new Random();

    const string sdkType = "statsig-server-core-dotnet";
    const string specNamePath = "/shared-volume/spec_names.json";
    const int ITER_LITE = 1000;
    const int ITER_HEAVY = 10_000; 


    static Dictionary<string, List<string>> LoadSpecNames()
    {
        for (int i = 0; i < 10; i++)
        {
            if (File.Exists(specNamePath))
                break;
            Task.Delay(1000).Wait();
        }
        var json = File.ReadAllText(specNamePath);
        return JsonSerializer.Deserialize<Dictionary<string, List<string>>>(json);
    }

    static StatsigCore::Statsig.StatsigUser CreateUser()
    {
        var user = new StatsigUserBuilder()
            .SetUserID("user_" + rng.Next(1_000_000))
            .SetCountry("US")
            .SetLocale("en-US")
            .SetAppVersion("1.0.0")
            .SetUserAgent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36")
            .AddCustomProperty("isAdmin", false)
            .AddPrivateAttribute("isPaid", "nah")
            .Build();
        return user;
    }

    static BenchmarkResult RunBenchmark(string benchName, string specName, int iterations, Action action)
    {
        var durations = new List<double>(iterations);
        var sw = new Stopwatch();
        for (int i = 0; i < iterations; i++)
        {
            sw.Restart();
            action();
            sw.Stop();
            durations.Add(sw.Elapsed.TotalMilliseconds);
        }
        
        durations.Sort();
        int p99Index = (int)(iterations * 0.99);
        
        var result = new BenchmarkResult
        {
            BenchmarkName = benchName,
            P99 = durations[Math.Min(p99Index, durations.Count - 1)],
            Max = durations.Last(),
            Min = durations.First(),
            Median = durations[durations.Count / 2],
            Avg = durations.Average(),
            SpecName = specName,
            SdkType = sdkType,
            SdkVersion = sdkVersion
        };


        Console.WriteLine($"{result.BenchmarkName.PadRight(30)} p99({result.P99.ToString("F4")}ms) max({result.Max.ToString("F4")}ms) {result.SpecName}");

        return result;
    }

    public static async Task Run()
    {
        Console.WriteLine($"Statsig .NET Core (v{sdkVersion})");
        Console.WriteLine("--------------------------------");

        var specNames = LoadSpecNames();

        var options = new StatsigCore::Statsig.StatsigOptionsBuilder().SetSpecsURL("http://scrapi:8000/v2/download_config_specs").SetLogEventURL("http://scrapi:8000/v1/log_event").Build();

        var statsig = new StatsigCore::Statsig.Statsig("secret-DOTNET_CORE", options);
        await statsig.Initialize();

        var results = new List<BenchmarkResult>();
        var globalUser = new StatsigCore::Statsig.StatsigUserBuilder().SetUserID("global_user").Build();

        // Feature gates
        foreach (var gate in specNames["feature_gates"])
        {
            // TODO(weihao): remove the delay
            // Rigth now, we want to make sure check gate works properly
            results.Add(RunBenchmark("check_gate", gate, ITER_HEAVY, () => statsig.CheckGate(CreateUser(), gate)));
            results.Add(RunBenchmark("check_gate_global_user", gate, ITER_HEAVY, () => statsig.CheckGate(globalUser, gate)));
            
            results.Add(RunBenchmark("get_feature_gate", gate, ITER_HEAVY, () => statsig.GetFeatureGate(CreateUser(), gate)));
            await Task.Delay(1);

            results.Add(RunBenchmark("get_feature_gate_global_user", gate, ITER_HEAVY, () => statsig.GetFeatureGate(globalUser, gate)));
            await Task.Delay(1);
        }

        // Dynamic configs
        foreach (var config in specNames["dynamic_configs"])
        {
            results.Add(RunBenchmark("get_dynamic_config", config, ITER_HEAVY, () => statsig.GetDynamicConfig(CreateUser(), config)));
            await Task.Delay(1);
     
            results.Add(RunBenchmark("get_dynamic_config_global_user", config, ITER_HEAVY, () => statsig.GetDynamicConfig(globalUser, config)));
            await Task.Delay(1);
        }

        // Experiments
        foreach (var experiment in specNames["experiments"])
        {
            results.Add(RunBenchmark("get_experiment", experiment, ITER_HEAVY, () => statsig.GetExperiment(CreateUser(), experiment)));
            await Task.Delay(1);
     
            results.Add(RunBenchmark("get_experiment_global_user", experiment, ITER_HEAVY, () => statsig.GetExperiment(globalUser, experiment)));
            await Task.Delay(1);
        }

        // Layers
        foreach (var layer in specNames["layers"])
        {
            results.Add(RunBenchmark("get_layer", layer, ITER_HEAVY, () => statsig.GetLayer(CreateUser(), layer)));
            await Task.Delay(1);
      
            results.Add(RunBenchmark("get_layer_global_user", layer, ITER_HEAVY, () => statsig.GetLayer(globalUser, layer)));
            await Task.Delay(1);
        }

        // Client initialize response (if available in .NET SDK)
        results.Add(RunBenchmark("get_client_initialize_response", "n/a", ITER_LITE, () => statsig.GetClientInitializeResponse(CreateUser())));
        await Task.Delay(1);
    
        results.Add(RunBenchmark("get_client_initialize_response_global_user", "n/a", ITER_LITE, () => statsig.GetClientInitializeResponse(globalUser)));
        await Task.Delay(1);

        await statsig.Shutdown();

        var resultsFile = $"/shared-volume/{sdkType}-{sdkVersion}-results.json";
        File.WriteAllText(resultsFile, JsonSerializer.Serialize(new
        {
            sdkType,
            sdkVersion,
            results
        }, new JsonSerializerOptions { WriteIndented = true }));
    }

    static string GetVersion()
    {
        var custom = typeof(StatsigCore::Statsig.Statsig).Assembly.CustomAttributes;

        foreach (var attribute in custom)
        {
            if (attribute.AttributeType != typeof(AssemblyInformationalVersionAttribute))
            {
                continue;
            }
        
            var value = attribute.ConstructorArguments[0].Value;
            var version = value?.ToString();
            if (version == null)
            {
                throw new Exception("Version is Null");
            }
            var parts = version.Split('+');
            return parts[0];
        } 
        
        throw new Exception("AssemblyInformationalVersionAttribute not Found");
    }
}

